public/login.php

```diff
@@ -177,62 +164,91 @@
         $httpUtils->addURLParameters(Module::getModuleURL('casserver/loggedIn.php'), $parameters),
     );
 }
+
+$defaultTicketName = isset($_GET['service']) ? 'ticket' : 'SAMLart';
+$ticketName = $casconfig->getOptionalValue('ticketName', $defaultTicketName);
+
+// Get the state.
+// If we come from an authproc filter, we will load the state from the stateId.
+// If not, we will get the state from the AuthSource Data
+try {
+    $state = $authProcId !== null ?
+        $attributeExtractor->manageState($authProcId) :
+        $as->getAuthDataArray();
+} catch (\SimpleSAML\Error\NoState $e) {
+    var_export($e, true);
+    die();
+}
+// Attribute Handler
+$state['ReturnTo'] = $returnUrl;
+if($authProcId !== null) {
+    $state[ProcessingChain::AUTHPARAM] = $authProcId;
+}
+try {
+    $mappedAttributes = $attributeExtractor->extractUserAndAttributes($state);
+} catch (\SimpleSAML\Error\Exception $e) {
+    var_export($e, true);
+    die();
+}
+
+$serviceTicket = $ticketFactory->createServiceTicket([
+                                                         'service' => $serviceUrl,
+                                                         'forceAuthn' => $forceAuthn,
+                                                         'userName' => $mappedAttributes['user'],
+                                                         'attributes' => $mappedAttributes['attributes'],
+                                                         'proxies' => [],
+                                                         'sessionId' => $sessionTicket['id'],
+                                                     ]);
+
+$ticketStore->addTicket($serviceTicket);
+
+$parameters[$ticketName] = $serviceTicket['id'];
+
+$validDebugModes = ['true', 'samlValidate'];
+
+// DEBUG MODE
+if (
+    array_key_exists('debugMode', $_GET) &&
+    in_array($_GET['debugMode'], $validDebugModes, true) &&
+    $casconfig->getOptionalBoolean('debugMode', false)
+) {
+    if ($_GET['debugMode'] === 'samlValidate') {
+        $samlValidate = new SamlValidateResponder();
+        $samlResponse = $samlValidate->convertToSaml($serviceTicket);
+        $soap = $samlValidate->wrapInSoap($samlResponse);
+        echo '<pre>' . htmlspecialchars(strval($soap)) . '</pre>';
+    } else {
+        $method = 'serviceValidate';
+        // Fake some options for validateTicket
+        $_GET[$ticketName] = $serviceTicket['id'];
+        // We want to capture the output from echo used in validateTicket
+        ob_start();
+        require_once 'utility/validateTicket.php';
+        $casResponse = ob_get_contents();
+        ob_end_clean();
+        echo '<pre>' . htmlspecialchars($casResponse) . '</pre>';
+    }
+} elseif ($redirect) {
+    // GET
+    $httpUtils->redirectTrustedURL($httpUtils->addURLParameters($serviceUrl, $parameters));
+} else {
+    // POST
+    try {
+        $httpUtils->submitPOSTData($serviceUrl, $parameters);
+    } catch (\SimpleSAML\Error\Exception $e) {
```

Comment:
curious why handle the exception this way instead of allowing it to be handled by SSP's exception handler? or is that just for while doing dev?

Reply:
> or is that just for while doing dev?

yes. This is going away. As soon as we confirm that the feature works i will further refine the code and it will go away.